internal/hcsoci/hcsdoc_wcow.go

```diff
@@ -80,27 +80,69 @@ func createWindowsContainerDocument(ctx context.Context, coi *createOptionsInter
-			// minroot configuration (host doesn't have access to all available LPs)
-			// so prefer standard OS level calls here instead.
-			hostCPUCount = processorinfo.ProcessorCount()
+			uvmCPUCount = coi.HostingSystem.ProcessorCount()
+			usableCPUCount = uvmCPUCount
 		}
-		if cpuCount > hostCPUCount {
+		if cpuCount > usableCPUCount {
 			l := log.G(ctx).WithField(logfields.ContainerID, coi.ID)
 			if coi.HostingSystem != nil {
 				l.Data[logfields.UVMID] = coi.HostingSystem.ID()
 			}
 			l.WithFields(logrus.Fields{
 				"requested": cpuCount,
-				"assigned":  hostCPUCount,
+				"assigned":  usableCPUCount,
 			}).Warn("Changing user requested CPUCount to current number of processors")
-			cpuCount = hostCPUCount
+			cpuCount = usableCPUCount
+		}
+		if coi.ScaleCPULimitsToSandbox && cpuLimit > 0 && coi.HostingSystem != nil {
+			// When ScaleCPULimitsToSandbox is set and we are running in a UVM, we assume
+			// the CPU limit has been calculated based on the number of processors on the
+			// host, and instead re-calculate it based on the number of processors in the UVM.
+			//
+			// This is needed to work correctly with assumptions kubelet makes when computing
+			// the CPU limit value:
+			// - kubelet thinks about CPU limits in terms of millicores, which are 1000ths of
+			//   cores. So if 2000 millicores are assigned, the container can use 2 processors.
+			// - In Windows, the job object CPU limit is global across all processors on the
+			//   system, and is represented as a fraction out of 10000. In this model, a limit
+			//   of 10000 means the container can use all processors fully, regardless of how
+			//   many processors exist on the system.
+			// - To convert the millicores value into the job object limit, kubelet divides
+			//   the millicores by the number of CPU cores on the host. This causes problems
+			//   when running inside a UVM, as the UVM may have a different number of processors
+			//   than the host system.
+			//
+			// To work around this, we undo the division by the number of host processors, and
+			// re-do the division based on the number of processors inside the UVM. This will
+			// give the correct value based on the actual number of millicores that the kubelet
+			// wants the container to have.
+			//
+			// Kubelet formula to compute CPU limit:
+			// cpuMaximum := 10000 * cpuLimit.MilliValue() / int64(runtime.NumCPU()) / 1000
+			newCPULimit := cpuLimit * hostCPUCount / uvmCPUCount
+			// We only apply bounds here because we are calculating the CPU limit ourselves,
+			// and this matches the kubelet behavior where they also bound the CPU limit by [1, 10000].
+			// In the case where we use the value directly from the user, we don't alter it to fit
+			// within the bounds, but just let the platform throw an error if it is invalid.
+			if newCPULimit < 1 {
+				newCPULimit = 1
+			} else if newCPULimit > 10000 {
+				newCPULimit = 10000
+			}
```

Comment:
is there a reason that kubelet bounds to this? Why can't we just let the platform fail here as well if it's invalid?

Reply:
Don't know why the kubelet chose to bound to it rather than just returning an error, but given that this code path is intended to work nicely with the kubelet, I feel we should keep the same behavior. For instance, if someone requests 1200 millicores on a 1-core system for a process-isolated container, kubelet will bound the limit at 10000, so we should do the same thing in the UVM scenario.

Reply:
I think that's reasonable now that I think through the scenario more.